bson-record-codec/src/main/org/bson/codecs/record/RecordCodec.java

```diff
@@ -0,0 +1,217 @@
+        this.clazz = notNull("class", clazz);
+        canonicalConstructor = notNull("canonicalConstructor", getCanonicalConstructor(clazz));
+        componentModels = getComponentModels(clazz, codecRegistry);
+        fieldNameToComponentModel = componentModels.stream()
+                .collect(Collectors.toMap(ComponentModel::getFieldName, Function.identity()));
+        componentModelForId = getComponentModelForId(clazz, componentModels);
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public T decode(final BsonReader reader, final DecoderContext decoderContext) {
+        reader.readStartDocument();
+
+        Object[] constructorArguments = new Object[componentModels.size()];
+        while (reader.readBsonType() != BsonType.END_OF_DOCUMENT) {
+            var fieldName = reader.readName();
+            var componentModel = fieldNameToComponentModel.get(fieldName);
+            constructorArguments[componentModel.index] = decoderContext.decodeWithChildContext(componentModel.codec, reader);
+        }
+        reader.readEndDocument();
+
+        try {
+            return (T) canonicalConstructor.newInstance(constructorArguments);
+        } catch (ReflectiveOperationException e) {
+            throw new CodecConfigurationException(format("Unable to invoke canonical constructor of record class %s", clazz.getName()), e);
+        }
+    }
+
+    @Override
+    public void encode(final BsonWriter writer, final T record, final EncoderContext encoderContext) {
+        writer.writeStartDocument();
+        if (componentModelForId != null) {
+            writeComponent(writer, record, componentModelForId);
+        }
+        for (var componentModel : componentModels) {
+            if (componentModel == componentModelForId) {
+                continue;
+            }
+            writeComponent(writer, record, componentModel);
+        }
+        writer.writeEndDocument();
+
+    }
+
+    @Override
+    public Class<T> getEncoderClass() {
+        return clazz;
+    }
+
+    @SuppressWarnings({"unchecked", "rawtypes"})
+    private void writeComponent(final BsonWriter writer, final T record, final ComponentModel componentModel) {
+        try {
+            Object componentValue = componentModel.getValue(record);
+            if (componentValue != null) {
+                writer.writeName(componentModel.getFieldName());
+                ((Codec) componentModel.codec).encode(writer, componentValue, EncoderContext.builder().build());
+            }
+        } catch (ReflectiveOperationException e) {
+            throw new CodecConfigurationException(
+                    format("Unable to access value of component %s for record %s", componentModel.getComponentName(), clazz.getName()), e);
+        }
+    }
+
+    private static <T> List<ComponentModel> getComponentModels(final Class<T> clazz, final CodecRegistry codecRegistry) {
+        var componentModels = new ArrayList<ComponentModel>(clazz.getRecordComponents().length);
```

Comment:
I don't think we care much about performance here because this is called once per record class, and not for every encode/decode.  But aesthetically it seems more pleasing to store in a local variable rather than call it three times.

Reply:
Done